rdmalib/include/rdmalib/buffer.hpp

```diff
@@ -2,192 +2,396 @@
+      _header(header),
+      _bytes(size * byte_size + header),
+      _byte_size(byte_size),
+      _mr(nullptr),
+      _own_memory(true)
+    {
+      // size_t alloc = _bytes;
+      // if(alloc < 4096) {
+      //   alloc = 4096;
+      //   spdlog::warn("Page too small, allocating {} bytes", alloc);
+      // }
+      //  page-aligned address for maximum performance
+      _ptr = mmap(nullptr, _bytes, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, 0, 0);
+      SPDLOG_DEBUG(
+          "Allocated {} bytes, address {}",
+          _bytes, fmt::ptr(_ptr));
+    }
+
+    template <typename Derived, typename Library>
+    Buffer<Derived, Library>::Buffer(void *ptr, uint32_t size, uint32_t byte_size) : _size(size),
+      _header(0),
+      _bytes(size * byte_size),
+      _byte_size(byte_size),
+      _ptr(ptr),
+      _mr(nullptr),
+      _own_memory(false)
+    {
+      SPDLOG_DEBUG(
+          "Allocated {} bytes, address {}",
+          _bytes, fmt::ptr(_ptr));
+    }
+    template <typename Derived, typename Library>
+    uint32_t Buffer<Derived, Library>::data_size() const
+    {
+      return this->_size;
+    }
+
+    template <typename Derived, typename Library>
+    uint32_t Buffer<Derived, Library>::size() const
+    {
+      return this->_size + this->_header;
+    }
+
+    template <typename Derived, typename Library>
+    uint32_t Buffer<Derived, Library>::bytes() const
+    {
+      return this->_bytes;
+    }
+    template <typename Derived, typename Library>
+    uintptr_t Buffer<Derived, Library>::address() const
+    {
+      assert(this->_mr);
+      return reinterpret_cast<uint64_t>(this->_ptr);
+    }
+
+    template <typename Derived, typename Library>
+    void *Buffer<Derived, Library>::ptr() const
+    {
+      return this->_ptr;
+    }
+
+  } /* end impl block */
+
+  template <typename Library>
+  struct RemoteBuffer
+  {
+    using rkey_t = typename library_traits<Library>::rkey_t;
 
-  struct RemoteBuffer {
     uintptr_t addr;
-    uint64_t rkey;
+    rkey_t rkey;
     uint32_t size;
 
-    RemoteBuffer();
+    RemoteBuffer() :
+      addr(0),
+      rkey(0),
+      size(0)
+    {
+    }
     // When accessing the remote buffer, we might not need to know the size.
-    RemoteBuffer(uintptr_t addr, uint64_t rkey, uint32_t size = 0);
-
-    #ifdef USE_LIBFABRIC
-    RemoteBuffer(uintptr_t addr, uint64_t rkey, uint32_t size);
-    #else
-    RemoteBuffer(uintptr_t addr, uint32_t rkey, uint32_t size);
-    #endif
+    RemoteBuffer(uintptr_t addr, rkey_t rkey, uint32_t size = 0) :
+      addr(addr),
+      rkey(rkey),
+      size(size)
+    {
+    }
 
-    template<class Archive>
-    void serialize(Archive & ar)
+    template <class Archive>
+    void serialize(Archive &ar)
     {
       ar(CEREAL_NVP(addr), CEREAL_NVP(rkey), CEREAL_NVP(size));
     }
   };
 
-  template<typename T>
-  struct Buffer : impl::Buffer{
+  struct LibfabricRemoteBuffer : RemoteBuffer<libfabric>
```

Comment:
Is there any difference between both implementations when it comes to `RemoteBufer`?